### PR TITLE
haskell_stack bugfixes, workarounds for macOS 11 issues

### PR DESCRIPTION
**Pull request by essandess**
# haskell_stack bugfixes, workarounds for macOS 11 issues

* Fixes: https://trac.macports.org/ticket/62228

Related:
* https://trac.macports.org/ticket/60477
* https://trac.macports.org/ticket/61422
* https://github.com/commercialhaskell/stack/issues/5456
* https://github.com/commercialhaskell/stack/issues/5491

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 11.2.1 20D74
Xcode 12.4 12D4e

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by Ionic** [comment]
Please split out the `stack_root` change into a separate commit, since it's really has its own semantics.

Drop https://trac.macports.org/ticket/60477 from the commit messages, since this PR is not actually touching this issue. I'm just waiting on macports/macports-legacy-support#34 to move and a `legacy-support` release to be done so that I can get my changes to the port in that fix this issue. Patience on that one, please.

Regarding "fixing" the issue on macOS 11 with a preloaded support library: urgh. I mean, it seems to work for you and the general idea isn't that weird, but I'd say it doesn't make sense to do this. Here's why:
  1. We'll have to remove the port again once stack switches to a new, working GHC version, which just adds weird noise.
  2. More importantly, aren't we eventually getting a stack binary that will fail when users execute it later on with this workaround? Every user would need to preload the library explicitly for operations via `stack` to work once the port is installed, but how would they know that? We could add notes (like in the `go` port for older systems), but those are often overlooked.

I'm not really in favor of this. I'd wait for stack's upstream to move and switch to the newer, working GHC version in order to fix that issue transparently. I know that stack seems to be slow, but that's really the better option. Preloading the library will just confuse users. IMHO, leaving it broken and fixing it with a stack upgrade transparently sounds like the most unsurprising way to go.

**Comment by essandess** [comment]
@Ionic Thanks for your comments. I will do what you suggest and restructure the `stack_root` fix into a separate PR.

I’m on the fence on merging the macOS 11 hacks. 

Personally, I became impatient getting these capabilities and their downstream dependencies installed on a macOS 11 box, so I went ahead and hacked them in on local ports. I suspect many others will just want them hacked in by now, and this PR accomplishes that.

Until the new stack binary is released, I don’t know how it will interact with `DYLD_INSERT_LIBRARIES=${prefix}/lib/macos11ghcwa.dylib`. I’m not aware of any reason it should break.

In any case, whenever the updated `stack` Is released, this block of code can be removed as part of the update, so there should be no issue.

**Comment by Ionic** [comment]
> Personally, I became impatient getting these capabilities and their downstream dependencies installed on a macOS 11 box, so I went ahead and hacked them in on local ports. I suspect many others will just want them hacked in by now, and this PR accomplishes that.

I can understand the impatience, given that stack's upstream is not... good on response times. It's still a bit of a surprise for users, let me explain why...

> Until the new stack binary is released, I don’t know how it will interact with DYLD_INSERT_LIBRARIES=${prefix}/lib/macos11ghcwa.dylib. I’m not aware of any reason it should break.

The problem is not that any port depending on stack in MacPorts will break. The problem is that we install a stack binary which will not work out of the box for users. It will work for our dependencies, since we use `DYLD_INSERT_LIBRARIES` in the PortGroup, but once a user tries to build its own project with stack, that'll fail, unless the user explicitly also uses `DYLD_INSERT_LIBRARIES`. That's the thing that gives me a bad feeling about the hack.

**Comment by essandess** [comment]
> once a user tries to build its own project with stack, that'll fail, unless the user explicitly also uses `DYLD_INSERT_LIBRARIES`. That's the thing that gives me a bad feeling about the hack.

Proposed solution in [force-push](https://github.com/macports/macports-ports/compare/974da7dde197310ee1175a0e9aa37f7aca8852ad..81796d358bfb6b5d158da219f6660799b25a06cd):
```
    notes           "On macOS 11 (Big Sur), a workaround to an issue \          
                    with previous versions of ghc require setting the \         
                    environment variable                                        
                                                                                
DYLD_INSERT_LIBRARIES=${prefix}/lib/macos11ghcwa.dylib                          
                                                                                
Reference: https://github.com/commercialhaskell/stack/issues/5456"
```

**Comment by Ionic** [comment]
Better use `notes-append` to not overwrite previous notes.

Yeah, that at least CAN alert users. A lot of them don't read those, but I guess that there's nothing more we could do, other than wait for stack's upstream to fix the issue by upgrading to a newer version of GHC. And that might take too long to be comfortable.

**Comment by Ionic** [comment]
Oh, also, since the library is also required at run time, it should be `depends-lib`.

**Comment by essandess** [comment]
See https://github.com/macports/macports-ports/pull/10089

**Comment by essandess** [comment]
I don't believe so. This is only to fix previous versions of `ghc` that `stack` uses for its build. The resulting binaries don't know about that:
```bash
otool -L `which happy`
/opt/local/bin/happy:
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/lib/libcharset.1.dylib (compatibility version 2.0.0, current version 2.0.0)
```

> Oh, also, since the library is also required at run time, it should be `depends-lib`.

**Comment by Ionic** [comment]
> See #10089

Oh, a separate commit as part of this PR would have sufficed, but that also works. :)

> I don't believe so. This is only to fix previous versions of ghc that stack uses for its build. The resulting binaries don't know about that: [...]

True, preloaded libraries will never show up.

But, as already mentioned, we want users to be able to build their own projects with stack, too. If we keep it a `build` dependency, `port_cutleaves -b` will rightfully remove it after the `stack` port has been installed (if users choose to execute it). That doesn't make sense and makes the note unuseful, too. If we make it a `lib` dependency, it will not be a leaf until stack (or any other dependent port) gets removed.

However, you do have a point in that the dependency is probably misplaced. It's not `happy` that needs the library - it's only stack. So maybe move that into the `stack` port directly with the type changed to `lib`?

**Comment by Ionic** [comment]
> However, you do have a point in that the dependency is probably misplaced. It's not `happy` that needs the library - it's only stack. So maybe move that into the `stack` port directly with the type changed to `lib`?

Unless, of course, packages like happy use their own, bundled binary copy of `stack` instead of the system version. In that case, it'll need to stay in the PG. I don't quite remember how this is set up currently.

**Comment by Ionic** [comment]
Hmm, out of curiosity, have you ever tried to build stack and dependent ports like happy with `haskell_stack.system_ghc yes` and the newer, fixed GHC version? Maybe we can completely avoid the hack, if we can just use a working GHC version (albeit newer, which may confuse stack's snapshot feature).

I'm not sure if `stack` is actually capable of bootstrapping with a system GHC version, but it may be worth a try.

**Comment by essandess** [comment]
That's a 💩🕳. The whole reason for `stack`'s existence is robust builds where they choose all the versions that will work. Second-guessing any of it opens the door to some deep sublevel of dependency hell.

That said, I bet it'd work, but time is finite.

> Hmm, out of curiosity, have you ever tried to build stack and dependent ports like happy with `haskell_stack.system_ghc yes` and the newer, fixed GHC version? Maybe we can completely avoid the hack, if we can just use a working GHC version (albeit newer, which may confuse stack's snapshot feature).
> 
> I'm not sure if `stack` is actually capable of bootstrapping with a system GHC version, but it may be worth a try.

**Comment by kencu** [comment]
> Hmm, out of curiosity, have you ever tried to build stack and dependent ports like happy with `haskell_stack.system_ghc yes` and the newer, fixed GHC version? Maybe we can completely avoid the hack, if we can just use a working GHC version (albeit newer, which may confuse stack's snapshot feature).
> 
> I'm not sure if `stack` is actually capable of bootstrapping with a system GHC version, but it may be worth a try.

Works perfectly.

https://github.com/kencu/ghc-for-older-darwin-systems/releases

This is for 10.6+

**Comment by Ionic** [comment]
> Second-guessing any of it opens the door to some deep sublevel of dependency hell.

I guess... I mean, in general, it shouldn't cause a lot of problems with dependencies, but it's conceivable, that some dependency is NOT building with a newer version of GHC, which would be a blocker. The preload hack is probably safer in that regard, but technically also changes ghc itself, which would make it incompatible with what stack assumes. But much less so than using a newer version, I guess.

Okay, a few more things, which I will comment about inline.

**Comment by essandess** [comment]
> > Second-guessing any of it opens the door to some deep sublevel of dependency hell.
> 
> in general, it shouldn't cause a lot of problems with dependencies, but it's conceivable, that some dependency is NOT building with a newer version of GHC, which would be a blocker.

FYI for stack itself, there’s this circularity:
```bash
port rdeps ghc | grep -B 1 stack
  alex
    stack
```

It can be hacked to work, but much, much better to go with a simpler workaround like `+prebuilt`.

**Review by Ionic — CHANGES_REQUESTED** [review]

**Comment by Ionic** [comment]
> FYI for stack itself, there’s this circularity:

Okay, yeah, circular dependencies are not supported and will have to be broken anyway. So that wouldn't have worked out anyway, I guess.

File: _resources/port1.0/group/haskell_stack-1.0.tcl [hunk 1]
```diff
@@ -88,9 +88,23 @@ options haskell_stack.yaml
 default haskell_stack.yaml  {${worksrcpath}/stack.yaml}
 
 options haskell_stack.env
-default haskell_stack.env \
-    {STACK_ROOT=[option haskell_stack.stack_root] \
-         STACK_YAML=[option haskell_stack.yaml]}
+
```
[the remark this answers is not in the document]
Comment (essandess) [reply _resources/port1.0/group/haskell_stack-1.0.tcl:94]: Thanks done.

File: _resources/port1.0/group/haskell_stack-1.0.tcl [hunk 2]
```diff
@@ -88,9 +88,23 @@ options haskell_stack.yaml
...
+# workaround for stack/ghc issue on macOS 11
+# https://github.com/commercialhaskell/stack/issues/5456
+# https://github.com/yairchu/macos11-haskell-workaround
+# remove after this issue is resolved
+if {[lindex [split ${macosx_deployment_target} .] 0] eq 11} {
```
[the remark this answers is not in the document]
Comment (essandess) [reply _resources/port1.0/group/haskell_stack-1.0.tcl]: Thanks done.

File: devel/macos11ghcwa/Portfile [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+maintainers         nomaintainer
+fetch.type          git
+
+description         Make Haskell work on macOS again!
+
+long_description    A work around for a bug in older GHCs (<= 8.10.3) \
+                    on macOS 11 "Big Sur" (bug #18446).
+
+use_configure       no
+
```
[the remark this answers is not in the document]
Comment (essandess) [reply devel/macos11ghcwa/Portfile:34]: Thanks done. This code doesn't throw an error, so I just used the `error` command.

File: devel/macos11ghcwa/Portfile [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+build {
+    system -W ${worksrcpath} \
+        "clang -target x86_64-darwin -dynamiclib macos11ghcwa.c -o macos11ghcwa.dylib"
```
[the remark this answers is not in the document]
Comment (essandess) [reply devel/macos11ghcwa/Portfile]: Thanks done.

This is simply and cut-and-paste of upstream build instructions:

https://github.com/yairchu/macos11-haskell-workaround/blob/62f08fc0ed61cc4c0fc8ac7c2dec558c85b4d4f3/README.md#L11

I don't know enough about this code to comment on flags for other architectures.

**Comment by kencu** [comment]
Hi fellows -- FYI here is GHC 9.0.1 for all Darwin systems 10 and greater:

<https://github.com/kencu/ghc-for-older-darwin-systems/releases/tag/9.0.1>

**Comment by essandess** [comment]
@neverpanic An update for `stack` hasn’t appeared yet and it’s March.

would you please weigh in on this workaround and if you agree merge the PR?

**Comment by neverpanic** [comment]
Not a fan of this, really. Using `DYLD_INSERT_LIBRARIES` and loading a library that interposes `stat` and `execve` will conflict with trace mode's own shenanigans that do this exact thing. I know that sometimes we backport fixes from upstream before they're willing to release them or even develop the fixes ourselves, but that does not usually happen in the form a a preload library that gets injected into binaries.

The dyld cache is a problem for all software running on macOS, and we were hit by it as well, and managed to fix this in a (somewhat) reasonable timeframe, despite our lack of developer time. I would much prefer if we could work with haskell stack to push a release out the door with a fix for this, or switch our version of GHC to what stack currently uses except with the required patch applied, and use our local GHC with the fix.

I'm not confident that patching around issues in a software that hasn't manged to get a release out the door with a bugfix between November and March is a sustainable strategy.

**Comment by essandess** [comment]
> that does not usually happen in the form a a preload library that gets injected into binaries.

Also, note that the library `libmacos11ghcwa` is not used by the resulting binaries that `stack` creates. It is used only in the build phase. See [above](https://github.com/macports/macports-ports/pull/10086#issuecomment-783716163). 
```bash
otool -L `which happy`
/opt/local/bin/happy:
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/lib/libcharset.1.dylib (compatibility version 2.0.0, current version 2.0.0)
```

**Comment by neverpanic** [comment]
>     * The whole point of `stack` is to download whatever version of `ghc` is specified by the LTS version to guarantee robust builds. If those `ghc` versions have this issue (most/all? likely do), then no build.
> 
>     * There is no guarantee that specifying the latest `ghc` version in which this issue is fixed will build existing projects. Maybe it will, but testing is a project in itself.

Note that I did not suggest using the latest GHC. I suggested using the particular version of GHC stack would have used by default (although I assume it can differ per port?), with a patch applied.


You have also not addressed my comments on what this will do when run in trace mode. My assumption is, it will just break.

**Comment by essandess** [comment]
> what this will do when run in trace mode. My assumption is, it will just break.

I believe that trace mode just breaks for existing `stack` ports anyway, without this macOS 11 workaround. For example, see the attached log of `port -vst destroot happy` done on 10.15: [main.log](https://github.com/macports/macports-ports/files/6075483/main.log).

**Comment by pmetzger** [comment]
Where are we on all of this? Updating `pandoc` is currently impossible...

**Comment by kencu** [comment]
> Where are we on all of this? Updating `pandoc` is currently impossible...

You could do what I did -- write your own pandoc portfile that installs the last binary and plan to come back to this question in a year or two...

**Comment by essandess** [comment]
> Where are we on all of this? Updating `pandoc` is currently impossible...

It is possible and straightforward. See #10514.

**Comment by pmetzger** [comment]
> It takes only a minute to write your own pandoc portfile that installs the last working binary.

That's not a solution for ordinary users who simply want to get their work done.

**Comment by kencu** [comment]
> > It takes only a minute to write your own pandoc portfile that installs the last working binary.
> 
> That's not a solution for ordinary users who simply want to get their work done.

It was, until it was removed from the Portfile :>

**Comment by kencu** [comment]
> > It takes only a minute to write your own pandoc portfile that installs the last working binary.
> 
> That's not a solution for ordinary users who simply want to get their work done.

Sometimes, it's the only practical thing to do. In general, it would be best if we would not break things for everyone while working out other solutions that have not  yet been accepted or finalized.

**Comment by kencu** [comment]
> > It takes only a minute to write your own pandoc portfile that installs the last working binary.
> 
> That's not a solution for ordinary users who simply want to get their work done.

Yes, well I fixed this once for everyone and it was stripped out without a viable backup plan in place, so at least I fixed my own. What else can I do?

**Comment by pmetzger** [comment]
What did you do that was removed? I am unaware of the history.

**Comment by kencu** [comment]
https://github.com/macports/macports-ports/commit/3ca592edfd14722597e29f8f18c8c781a5b52b56#diff-c925442da5b6f614ff77da74d25253ccb45e29e1b855fdc91ab58ed3c4ede6be

**Comment by pmetzger** [comment]
That doesn't tell me what happened, it only shows me the commit?

**Comment by kencu** [comment]
I always find out what actually happened in a port by using the commit history and git-blame. Everything else is ... not what actually happened :)

**Comment by neverpanic** [comment]
Actually, looking at this commit I'm wondering why we don't revert it. It's commit message says it removes a workaround that's obsoleted by this PR, but this PR isn't merged, and I still think it's a hack.

**Comment by essandess** [comment]
> Actually, looking at this commit I'm wondering why we don't revert it. It's commit message says it removes a workaround that's obsoleted by this PR, but this PR isn't merged, and I still think it's a hack.

The `pandoc` build issue on macOS 11 can be fixed by switching to a `cabal` build in:
* #10514
* #10601

I’ll defer to @neverpanic and others, but the tradeoff of explicitly relying on MacPorts binaries in the previous commit looks comparable—and more complicated—than implicitly relying on MacPorts binaries for `alex` to build the required `ghc`.

**Comment by essandess** [comment]
> > Actually, looking at this commit I'm wondering why we don't revert it. It's commit message says it removes a workaround that's obsoleted by this PR, but this PR isn't merged, and I still think it's a hack.
> 
> The `pandoc` build issue on macOS 11 can be fixed by switching to a `cabal` build in:
> 
> * #10514
> * #10601
> 
> I’ll defer to @neverpanic and others, but the tradeoff of explicitly relying on MacPorts binaries in the previous commit looks comparable—and more complicated—than implicitly relying on MacPorts binaries for `alex` to build the required `ghc`.

Also the `ghc` bugfix in #10596 is required.

**Comment by essandess** [comment]
> It's commit message says it removes a workaround that's obsoleted by this PR

This `pandoc` commit isn’t related to this PR. It’s a simple update: https://github.com/macports/macports-ports/pull/10514/commits/3d4a4145002bf11d5d8f846a4ccd536d841ebb6f
> pandoc: Update to version 2.13, move to PG haskell_cabal
> * Update to version 2.13
> * See:
>     * https://trac.macports.org/ticket/61422
>     * https://trac.macports.org/ticket/62579
> 
> Depends on #10601

**Comment by essandess** [comment]
Superseded by #10789.

**Pull request closed without merging**